test/regression/RegressionTest.cc

```diff
@@ -53,23 +50,29 @@ void RegressionTest::run(const char* source, const char* triple) {
       simeng::span<char>(reinterpret_cast<char*>(code_), codeSize_), config);
   ASSERT_TRUE(process_->isValid());
   uint64_t entryPoint = process_->getEntryPoint();
-
-  // Allocate memory for the process and copy the full process image to it
-  simeng::span<char> processImage = process_->getProcessImage();
-  processMemorySize_ = processImage.size();
-  if (processMemory_) delete[] processMemory_;
-  processMemory_ = new char[processMemorySize_];
-  std::copy(processImage.begin(), processImage.end(), processMemory_);
-
-  // Create memory interfaces for instruction and data access
-  simeng::FlatMemoryInterface instructionMemory(processMemory_,
+  processMemorySize_ = process_->getProcessImageSize();
+  // This instance of procImgPtr pointer needs to be shared because
+  // getMemoryValue in RegressionTest.hh uses reference to the class
+  // member processMemory_.
+  std::shared_ptr<char> procImgPtr = process_->getProcessImage();
+  processMemory_ = procImgPtr.get();
+
+  // Create memory interfaces for instruction and data access.
+  // For each memory interface, a dereferenced shared_ptr to the
+  // processImage is passed as argument.
+  std::shared_ptr<char> procImgForInstrMem = process_->getProcessImage();
+  simeng::FlatMemoryInterface instructionMemory(procImgForInstrMem.get(),
                                                 processMemorySize_);
+
+  std::shared_ptr<char> procImgForFlatDataMem = process_->getProcessImage();
   std::unique_ptr<simeng::FlatMemoryInterface> flatDataMemory =
-      std::make_unique<simeng::FlatMemoryInterface>(processMemory_,
+      std::make_unique<simeng::FlatMemoryInterface>(procImgForFlatDataMem.get(),
                                                     processMemorySize_);
+
+  std::shared_ptr<char> procImgForFixedDataMem = process_->getProcessImage();
```

Comment:
Rather than having 3 shared pointers which all point to the same process image, why can't we use a single `processImage` shared pointer and pass this to each MemoryInterface constructor?

Reply:
Thanks for pointing this out, I was confused with shared_ptr ownership so I copied it 3 times, which shouldn't be the case. Since the scope which contains these objects retains the ownership. I have fixed it now.

Reply:
PTAL